## Solve these exercises in the codeblock below each question and compare your answer

1.Import numpy as `np` and print the version number

In [7]:
import numpy as np
print(np.__version__)

1.26.2


1.19.5

2.Create a 1D array of numbers from 0 to 9 Desired output

In [13]:
#help(np.linspace)
a=np.linspace(0,9,10,dtype=int)
a

array([0, 1, 2, 3, 4, 5, 6, 7, 8, 9])

array([0, 1, 2, 3, 4, 5, 6, 7, 8, 9])

3.Create a 3×3 numpy array of all True’s

In [18]:
a = np.ones((3,3),dtype=bool)
a

array([[ True,  True,  True],
       [ True,  True,  True],
       [ True,  True,  True]])

array([[ True,  True,  True],</br>
&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;       [ True,  True,  True],</br>
       &nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp; [ True,  True,  True]])

4.Extract all odd numbers from arr Input

In [27]:
a = np.array([0, 1, 2, 3, 4, 5, 6, 7, 8, 9])
b = a[a%2 != 0]
b


array([1, 3, 5, 7, 9])

5. Replace all odd numbers in arr with -1 Input: <br>
arr = np.array([0, 1, 2, 3, 4, 5, 6, 7, 8, 9])

In [28]:
a[a%2 !=0 ] = -1
a

array([ 0, -1,  2, -1,  4, -1,  6, -1,  8, -1])

array([ 0, -1,  2, -1,  4, -1,  6, -1,  8, -1])

6.Replace all odd numbers in arr with -1 without changing arr Input:

arr = np.array([0, 1, 2, 3, 4, 5, 6, 7, 8, 9])

out<br>
array([ 0, -1,  2, -1,  4, -1,  6, -1,  8, -1])<br>
arr<br>
array([0, 1, 2, 3, 4, 5, 6, 7, 8, 9])

7.Convert a 1D array to a 2D array with 2 rows Input

np.arange(10)

array([ 0, 1, 2, 3, 4, 5, 6, 7, 8, 9])

In [33]:
a=np.arange(10)
b = a.reshape(2,5)
b

array([[0, 1, 2, 3, 4],
       [5, 6, 7, 8, 9]])

 array([[0, 1, 2, 3, 4],<br>
&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;        [5, 6, 7, 8, 9]])

8.Stack arrays a and b vertically Input

a = np.arange(10).reshape(2,-1)<br>
b = np.repeat(1, 10).reshape(2,-1)

In [60]:
a = np.arange(10).reshape(2,-1)
b = np.repeat(1, 10).reshape(2,-1)
print(a)
print(b) 
c = np.vstack((a,b))
print(c)

[[0 1 2 3 4]
 [5 6 7 8 9]]
[[1 1 1 1 1]
 [1 1 1 1 1]]
[[0 1 2 3 4]
 [5 6 7 8 9]
 [1 1 1 1 1]
 [1 1 1 1 1]]


array([[0, 1, 2, 3, 4],<br>
&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;        [5, 6, 7, 8, 9],<br>
&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;        [1, 1, 1, 1, 1],<br>
&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;        [1, 1, 1, 1, 1]])

9.Stack the arrays a and b horizontally. Input

a = np.arange(10).reshape(2,-1)

b = np.repeat(1, 10).reshape(2,-1)

In [61]:
a = np.arange(10).reshape(2,-1)

b = np.repeat(1, 10).reshape(2,-1)

print(a)
print(b) 
c = np.hstack((a,b))
print(c)

[[0 1 2 3 4]
 [5 6 7 8 9]]
[[1 1 1 1 1]
 [1 1 1 1 1]]
[[0 1 2 3 4 1 1 1 1 1]
 [5 6 7 8 9 1 1 1 1 1]]


array([[0, 1, 2, 3, 4, 1, 1, 1, 1, 1],<br>
&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;        [5, 6, 7, 8, 9, 1, 1, 1, 1, 1]])

10.Create the following pattern without hardcoding. Use only numpy functions and the below input array a.

 Input:

a = np.array([1,2,3])

In [80]:
a = np.array([1,2,3])
b = np.vstack((a,a,a)).transpose().reshape(9)
c = np.hstack((b,a,a,a))
c


array([1, 1, 1, 2, 2, 2, 3, 3, 3, 1, 2, 3, 1, 2, 3, 1, 2, 3])

array([1, 1, 1, 2, 2, 2, 3, 3, 3, 1, 2, 3, 1, 2, 3, 1, 2, 3])

11.Get the common items between a and b Input:

a = np.array([1,2,3,2,3,4,3,4,5,6])

b = np.array([7,2,10,2,7,4,9,4,9,8])

In [92]:
a = np.array([1,3,2,2,3,4,3,4,5,9])

b = np.array([7,2,10,2,7,4,9,4,9,8])

c = np.intersect1d(a,b)

c

array([2, 4, 9])

array([2, 4])

12.From array a remove all items present in array b Input:

a = np.array([1,2,3,4,5])

b = np.array([5,6,7,8,9])

In [104]:
a = np.array([1,2,3,4,5])
a = set(a)
b = np.array([5,6,7,8,9])
b = set(b)
c = a.intersection(b)

a.difference_update(c)
a = np.array(list(a))
a

array([1, 2, 3, 4])

array([1,2,3,4])

13.Get the positions where elements of a and b match Input:

a = np.array([1,2,3,2,3,4,3,4,5,6])

b = np.array([7,2,10,2,7,4,9,4,9,8])

In [115]:
a = np.array([1,2,3,2,3,4,3,4,5,6])

b = np.array([7,2,10,2,7,4,9,4,9,8])

c = np.intersect1d(a,b)

d = np.where((a==c[0]) | (a==c[1]))

d

(array([1, 3, 5, 7]),)

(array([1, 3, 5, 7]),)

14.Get all items between 5 and 10 from a.

Input:

a = np.array([2, 6, 1, 9, 10, 3, 27])

In [116]:
a = np.array([2, 6, 1, 9, 10, 3, 27])

b = a[(a>=5) & (a<=10)]

b

array([ 6,  9, 10])

(array([6, 9, 10]),)

 15.Convert the function maxx that works on two scalars, to work on two arrays.

 Input:

 def maxx(x, y):

    """Get the maximum of two items"""

    if x >= y:

        return x

    else:

        return y


maxx(1, 5)

> 5


a = np.array([5, 7, 9, 8, 6, 4, 5])

b = np.array([6, 3, 4, 8, 9, 7, 1])

pair_max(a, b)

array([ 6.,  7.,  9.,  8.,  9.,  7.,  5.])

16.Swap columns 1 and 2 in the array arr.

arr = np.arange(9).reshape(3,3)

arr

> array([[1, 0, 2],
>        [4, 3, 5],
>        [7, 6, 8]])

17.Swap rows 1 and 2 in the array arr:

arr = np.arange(9).reshape(3,3)


> array([[3, 4, 5],<br>
>        [0, 1, 2],<br>
>        [6, 7, 8]])

18.Reverse the rows of a 2D array arr.

Input

arr = np.arange(9).reshape(3,3)

array([[6, 7, 8],
       [3, 4, 5],
       [0, 1, 2]])

19.Reverse the columns of a 2D array arr.

 Input

arr = np.arange(9).reshape(3,3)

> array([[2, 1, 0],
>        [5, 4, 3],
>        [8, 7, 6]])

20.Create a 2D array of shape 5×3 to contain random decimal numbers between 5 and 10

> [[ 8.50061025  9.10531502  6.85867783]

>  [ 9.76262069  9.87717411  7.13466701]

>  [ 7.48966403  8.33409158  6.16808631]

>  [ 7.75010551  9.94535696  5.27373226]

>  [ 8.0850361   5.56165518  7.31244004]]

21. Print or show only 3 decimal places of the numpy array

 Input:

rand_arr = np.random.random((5,3))

> array([[ 0.443,  0.109,  0.97 ],
>        [ 0.388,  0.447,  0.191],
>        [ 0.891,  0.474,  0.212],
>        [ 0.609,  0.518,  0.403]])

22.Pretty print rand_arr by suppressing the scientific notation (like 1e10)

 Input:

 np.random.seed(100)

rand_arr = np.random.random([3,3])/1e3

rand_arr

 array([[  5.434049e-04,   2.783694e-04,   4.245176e-04],

        [  8.447761e-04,   4.718856e-06,   1.215691e-04],

        [  6.707491e-04,   8.258528e-04,   1.367066e-04]])


 array([[ 0.000543,  0.000278,  0.000425],<br>
        [ 0.000845,  0.000005,  0.000122],<br>
        [ 0.000671,  0.000826,  0.000137]])

23.Limit the number of items printed in python numpy array a to a maximum of 6 elements.

Input:

a = np.arange(15)
> array([ 0,  1,  2,  3,  4,  5,  6,  7,  8,  9, 10, 11, 12, 13, 14])

> array([ 0,  1,  2, ..., 12, 13, 14])

24.Print the full numpy array a without truncating.

 Input:

np.set_printoptions(threshold=6)

a = np.arange(15)
a

> array([ 0,  1,  2, ..., 12, 13, 14])

a
> array([ 0,  1,  2,  3,  4,  5,  6,  7,  8,  9, 10, 11, 12, 13, 14])

25.Import the iris dataset keeping the text intact.

Print the first 3 rows

iris[:3]

> array([[b'5.1', b'3.5', b'1.4', b'0.2', b'Iris-setosa'],<br>
>        [b'4.9', b'3.0', b'1.4', b'0.2', b'Iris-setosa'],<br>
>        [b'4.7', b'3.2', b'1.3', b'0.2', b'Iris-setosa']], dtype=object)

26.Extract the text column species from the 1D iris imported in previous question.

Input:

url = 'https://archive.ics.uci.edu/ml/machine-learning-databases/iris/iris.data'

 iris_1d = np.genfromtxt(url, delimiter=',', dtype=None)

species[:5]

> array([b'Iris-setosa', b'Iris-setosa', b'Iris-setosa', b'Iris-setosa',  b'Iris-setosa'],
    dtype='|S18')

27.Convert the 1D iris to 2D array iris_2d by omitting the species text field.

Input:

url = 'https://archive.ics.uci.edu/ml/machine-learning-databases/iris/iris.data'
iris_1d = np.genfromtxt(url, delimiter=',', dtype=None)

iris_2d[:4]
> array([[ 5.1,  3.5,  1.4,  0.2],<br>
>        [ 4.9,  3. ,  1.4,  0.2],<br>
>        [ 4.7,  3.2,  1.3,  0.2],<br>
>        [ 4.6,  3.1,  1.5,  0.2]])

28.Find the mean, median, standard deviation of iris’s sepallength (1st column)

url = 'https://archive.ics.uci.edu/ml/machine-learning-databases/iris/iris.data'
iris = np.genfromtxt(url, delimiter=',', dtype='object')

print(mu, med, sd)
> 5.84333333333 5.8 0.825301291785

29.Create a normalized form of iris‘s sepallength whose values range exactly between 0 and 1 so that the minimum has value 0 and maximum has value 1.

Input:

url = 'https://archive.ics.uci.edu/ml/machine-learning-databases/iris/iris.data'
sepallength = np.genfromtxt(url, delimiter=',', dtype='float', usecols=[0])

print(S)
> [ 0.222  0.167  0.111  0.083  0.194  0.306  0.083  0.194  0.028  0.167<br>
>   0.306  0.139  0.139  0.     0.417  0.389  0.306  0.222  0.389  0.222<br>
>   0.306  0.222  0.083  0.222  0.139  0.194  0.194  0.25   0.25   0.111<br>
>   0.139  0.306  0.25   0.333  0.167  0.194  0.333  0.167  0.028  0.222<br>
>   0.194  0.056  0.028  0.194  0.222  0.139  0.222  0.083  0.278  0.194<br>
>   0.75   0.583  0.722  0.333  0.611  0.389  0.556  0.167  0.639  0.25<br>
>   0.194  0.444  0.472  0.5    0.361  0.667  0.361  0.417  0.528  0.361<br>
>   0.444  0.5    0.556  0.5    0.583  0.639  0.694  0.667  0.472  0.389<br>
>   0.333  0.333  0.417  0.472  0.306  0.472  0.667  0.556  0.361  0.333<br>
>   0.333  0.5    0.417  0.194  0.361  0.389  0.389  0.528  0.222  0.389<br>
>   0.556  0.417  0.778  0.556  0.611  0.917  0.167  0.833  0.667  0.806<br>
>   0.611  0.583  0.694  0.389  0.417  0.583  0.611  0.944  0.944  0.472<br>
>   0.722  0.361  0.944  0.556  0.667  0.806  0.528  0.5    0.583  0.806<br>
>   0.861  1.     0.583  0.556  0.5    0.944  0.556  0.583  0.472  0.722<br>
>   0.667  0.722  0.417  0.694  0.667  0.667  0.556  0.611  0.528  0.444]<br>

30.Compute the softmax score of sepallength.

url = 'https://archive.ics.uci.edu/ml/machine-learning-databases/iris/iris.data'
sepallength = np.genfromtxt(url, delimiter=',', dtype='float', usecols=[0])

array([ 4.6  ,  7.255])

31Find the 5th and 95th percentile of iris’s sepallength

url = 'https://archive.ics.uci.edu/ml/machine-learning-databases/iris/iris.data'


sepallength = np.genfromtxt(url, delimiter=',', dtype='float', usecols=[0])

array([ 4.6  ,  7.255])

32.Insert np.nan values at 20 random positions in iris_2d dataset

 Input<br>
url = 'https://archive.ics.uci.edu/ml/machine-learning-databases/iris/iris.data'
iris_2d = np.genfromtxt(url, delimiter=',', dtype='object')

print(iris_2d[:10])
> [[b'5.1' b'3.5' b'1.4' b'0.2' b'Iris-setosa']<br>
>  [b'4.9' b'3.0' b'1.4' b'0.2' b'Iris-setosa']<br>
>  [b'4.7' b'3.2' b'1.3' b'0.2' b'Iris-setosa']<br>
>  [b'4.6' b'3.1' b'1.5' b'0.2' b'Iris-setosa']<br>
>  [b'5.0' b'3.6' b'1.4' b'0.2' b'Iris-setosa']<br>
>  [b'5.4' b'3.9' b'1.7' b'0.4' b'Iris-setosa']<br>
>  [b'4.6' b'3.4' b'1.4' b'0.3' b'Iris-setosa']<br>
>  [b'5.0' b'3.4' b'1.5' b'0.2' b'Iris-setosa']<br>
>  [b'4.4' nan b'1.4' b'0.2' b'Iris-setosa']<br>
>  [b'4.9' b'3.1' b'1.5' b'0.1' b'Iris-setosa']]

33.Find the number and position of missing values in iris_2d‘s sepallength (1st column)

 Input<br>
url = 'https://archive.ics.uci.edu/ml/machine-learning-databases/iris/iris.data'


iris_2d = np.genfromtxt(url, delimiter=',', dtype='float')

iris_2d[np.random.randint(150, size=20), np.random.randint(4, size=20)] =


 np.nan

(array([ 39,  88,  99, 130, 147]),)

34.Filter the rows of iris_2d that has petallength (3rd column) > 1.5 and sepallength (1st column) < 5.0

 Input<br>
url = 'https://archive.ics.uci.edu/ml/machine-learning-databases/iris/iris.data'<br>
iris_2d = np.genfromtxt(url, delimiter=',', dtype='float', usecols=[0
,1,2,3])

> array([[ 4.8,  3.4,  1.6,  0.2],<br>
>        [ 4.8,  3.4,  1.9,  0.2],<br>
>        [ 4.7,  3.2,  1.6,  0.2],<br>
>        [ 4.8,  3.1,  1.6,  0.2],<br>
>        [ 4.9,  2.4,  3.3,  1. ],<br>
>        [ 4.9,  2.5,  4.5,  1.7]])

35.Select the rows of iris_2d that does not have any nan value.

# Input
url = 'https://archive.ics.uci.edu/ml/machine-learning-databases/iris/iris.data'

iris_2d = np.genfromtxt(url, delimiter=',', dtype='float', usecols=[0,1,2,
3])

>array([[ 4.9,  3. ,  1.4,  0.2],<br>
>        [ 4.7,  3.2,  1.3,  0.2],<br>
>        [ 4.6,  3.1,  1.5,  0.2],<br>
>        [ 5. ,  3.6,  1.4,  0.2],<br>
>        [ 5.4,  3.9,  1.7,  0.4]])

36.Find the correlation between SepalLength(1st column) and PetalLength(3rd column) in iris_2d

 Input<br>
url = 'https://archive.ics.uci.edu/ml/machine-learning-databases/iris/iris.data'


iris_2d = np.genfromtxt(url, delimiter=',', dtype='float', usecols=[0,1,2,3])

> 0.871754157305

37.Find out if iris_2d has any missing values.

Input<br>
url = 'https://archive.ics.uci.edu/ml/machine-learning-databases/iris/iris.data'


iris_2d = np.genfromtxt(url, delimiter=',', dtype='float', usecols=[0,1,2,3])

FALSE

38.Replace all ccurrences of nan with 0 in numpy array

url = 'https://archive.ics.uci.edu/ml/machine-learning-databases/iris/iris.data'

iris_2d = np.genfromtxt(url, delimiter=',', dtype='float', usecols=[0,1,2,3])

iris_2d[np.random.randint(150, size=20), np.random.randint(4, size=20)] = np.nan

> array([[ 5.1,  3.5,  1.4,  0. ],<br>
>        [ 4.9,  3. ,  1.4,  0.2],<br>
 >     [ 4.7,  3.2,  1.3,  0.2],<br>
>        [ 4.6,  3.1,  1.5,  0.2]])<br>

39. Find the unique values and the count of unique values in iris’s species

 Input<br>
url = 'https://archive.ics.uci.edu/ml/machine-learning-databases/iris/iris.data'

iris = np.genfromtxt(url, delimiter=',', dtype='object')

names = ('sepallength', 'sepalwidth', 'petallength', 'petalwidth', 'species')

> (array([b'Iris-setosa', b'Iris-versicolor', b'Iris-virginica'],
>        dtype='|S15'), array([50, 50, 50]))

40.Difficulty Level: L2 Q. Bin the petal length (3rd) column of iris_2d to form a text array, such that if petal length is:

    Less than 3 –> ‘small’
    3-5 –> ‘medium’
    ‘>=5 –> ‘large’

 Input<br>
url = 'https://archive.ics.uci.edu/ml/machine-learning-databases/iris/iris.data'

iris = np.genfromtxt(url, delimiter=',', dtype='object')

names = ('sepallength', 'sepalwidth', 'petallength', 'petalwidth', 'species')

<#> ['small', 'small', 'small', 'small']

41.Create a new column for volume in iris_2d, where volume is (pi x petallength x sepal_length^2)/3

 Input<br>
url = 'https://archive.ics.uci.edu/ml/machine-learning-databases/iris/iris.data'

iris_2d = np.genfromtxt(url, delimiter=',', dtype='object')

names = ('sepallength', 'sepalwidth', 'petallength', 'petalwidth', 'species')

out[:4]
> array([[b'5.1', b'3.5', b'1.4', b'0.2', b'Iris-setosa', 38.13265162927291],<br>
>        [b'4.9', b'3.0', b'1.4', b'0.2', b'Iris-setosa', 35.200498485922445],<br>
>        [b'4.7', b'3.2', b'1.3', b'0.2', b'Iris-setosa', 30.0723720777127],<br>
>        [b'4.6', b'3.1', b'1.5', b'0.2', b'Iris-setosa', 33.238050274980004]], dtype=object)

42.Randomly sample iris‘s species such that setose is twice the number of versicolor and virginica

Import iris keeping the text column intact<br>
url = 'https://archive.ics.uci.edu/ml/machine-learning-databases/iris/iris.data'

iris = np.genfromtxt(url, delimiter=',', dtype='object')

> (array([b'Iris-setosa', b'Iris-versicolor', b'Iris-virginica'], dtype=object), array([77, 37, 36]))

43.What is the value of second longest petallength of species setosa

Input<br>
url = 'https://archive.ics.uci.edu/ml/machine-learning-databases/iris/iris.data'

iris = np.genfromtxt(url, delimiter=',', dtype='object')

names = ('sepallength', 'sepalwidth', 'petallength', 'petalwidth', 'species')

> 1.7

44.Sort the iris dataset based on sepallength column.

url = 'https://archive.ics.uci.edu/ml/machine-learning-databases/iris/iris.data'

iris = np.genfromtxt(url, delimiter=',', dtype='object')

names = ('sepallength', 'sepalwidth', 'petallength', 'petalwidth', 'species')

> [[b'4.3' b'3.0' b'1.1' b'0.1' b'Iris-setosa']<br>
>  [b'4.4' b'3.2' b'1.3' b'0.2' b'Iris-setosa']<br>
>  [b'4.4' b'3.0' b'1.3' b'0.2' b'Iris-setosa']<br>
>  [b'4.4' b'2.9' b'1.4' b'0.2' b'Iris-setosa']<br>
>  [b'4.5' b'2.3' b'1.3' b'0.3' b'Iris-setosa']<br>
>  [b'4.6' b'3.6' b'1.0' b'0.2' b'Iris-setosa']<br>
>  [b'4.6' b'3.1' b'1.5' b'0.2' b'Iris-setosa']<br>
>  [b'4.6' b'3.4' b'1.4' b'0.3' b'Iris-setosa']<br>
>  [b'4.6' b'3.2' b'1.4' b'0.2' b'Iris-setosa']<br>
>  [b'4.7' b'3.2' b'1.3' b'0.2' b'Iris-setosa']<br>
>  [b'4.7' b'3.2' b'1.6' b'0.2' b'Iris-setosa']<br>
>  [b'4.8' b'3.0' b'1.4' b'0.1' b'Iris-setosa']<br>
>  [b'4.8' b'3.0' b'1.4' b'0.3' b'Iris-setosa']<br>
>  [b'4.8' b'3.4' b'1.9' b'0.2' b'Iris-setosa']<br>
>  [b'4.8' b'3.4' b'1.6' b'0.2' b'Iris-setosa']<br>
>  [b'4.8' b'3.1' b'1.6' b'0.2' b'Iris-setosa']<br>
>  [b'4.9' b'2.4' b'3.3' b'1.0' b'Iris-versicolor']<br>
>  [b'4.9' b'2.5' b'4.5' b'1.7' b'Iris-virginica']<br>
>  [b'4.9' b'3.1' b'1.5' b'0.1' b'Iris-setosa']<br>
>  [b'4.9' b'3.1' b'1.5' b'0.1' b'Iris-setosa']]<br>

45.Find the most frequent value of petal length (3rd column) in iris dataset. Input:

url = 'https://archive.ics.uci.edu/ml/machine-learning-databases/iris/iris.data'

iris = np.genfromtxt(url, delimiter=',', dtype='object')

names = ('sepallength', 'sepalwidth', 'petallength', 'petalwidth', 'species')

> b'1.5'

46.Find the position of the first occurrence of a value greater than 1.0 in petalwidth 4th column of iris dataset.

 Input:

url = 'https://archive.ics.uci.edu/ml/machine-learning-databases/iris/iris.data'

iris = np.genfromtxt(url, delimiter=',', dtype='object')

> 50

47.From the array a, replace all values greater than 30 to 30 and less than 10 to 10.

Input:

np.random.seed(100)

a = np.random.uniform(1,50, 20)

> [ 27.63 &nbsp; 14.64 &nbsp; 21.8  &nbsp; 30. &nbsp;   10. &nbsp;   10. &nbsp;   30. &nbsp;   30. &nbsp;   10. &nbsp;   29.18 &nbsp; 30.
>   11.25<br> &nbsp; 10.08 &nbsp; 10. &nbsp;   11.77 &nbsp; 30. &nbsp;   30. &nbsp;   10. &nbsp;   30.  &nbsp;  14.43]

48.Get the positions of top 5 maximum values in a given array a.

np.random.seed(100)

a = np.random.uniform(1,50, 20)

49.Compute the counts of unique values row-wise. Input:

np.random.seed(100)
arr = np.random.randint(1,11,size=(6, 10))
arr
> array([[ 9,  9,  4,  8,  8,  1,  5,  3,  6,  3],<br>
>        [ 3,  3,  2,  1,  9,  5,  1, 10,  7,  3],<br>
>        [ 5,  2,  6,  4,  5,  5,  4,  8,  2,  2],<br>
>        [ 8,  8,  1,  3, 10, 10,  4,  3,  6,  9],<br>
>        [ 2,  1,  8,  7,  3,  1,  9,  3,  6,  2],<br>
>        [ 9,  2,  6,  5,  3,  9,  4,  6,  1, 10]])

> [[1, 0, 2, 1, 1, 1, 0, 2, 2, 0],<br>
>  [2, 1, 3, 0, 1, 0, 1, 0, 1, 1],<br>
>  [0, 3, 0, 2, 3, 1, 0, 1, 0, 0],<br>
>  [1, 0, 2, 1, 0, 1, 0, 2, 1, 2],<br>
>  [2, 2, 2, 0, 0, 1, 1, 1, 1, 0],<br>
>  [1, 1, 1, 1, 1, 2, 0, 0, 2, 1]]

50.Convert array_of_arrays into a flat linear 1d array. Input:

 Input:<br>
arr1 = np.arange(3)

arr2 = np.arange(3,7)

arr3 = np.arange(7,10)

array_of_arrays = np.array([arr1, arr2, arr3])

array_of_arrays
> array([array([0, 1, 2]), array([3, 4, 5, 6]), array([7, 8, 9])], dtype=object)

> array([0, 1, 2, 3, 4, 5, 6, 7, 8, 9])

51.Compute the one-hot encodings (dummy binary variables for each unique value in the array) Input:

np.random.seed(101) <br>
arr = np.random.randint(1,4, size=6)<br>
arr<br>
> array([2, 3, 2, 2, 2, 1])

> array([[ 0.,  1.,  0.],<br>
>        [ 0.,  0.,  1.],<br>
>        [ 0.,  1.,  0.],<br>
>        [ 0.,  1.,  0.],<br>
>        [ 0.,  1.,  0.],<br>
>        [ 1.,  0.,  0.]])

52.Create row numbers grouped by a categorical variable. Use the following sample from iris species as input. Input:

url = 'https://archive.ics.uci.edu/ml/machine-learning-databases/iris/iris.data'

species = np.genfromtxt(url, delimiter=',', dtype='str', usecols=4)

species_small = np.sort(np.random.choice(species, size=20))

species_small
> array(['Iris-setosa', 'Iris-setosa', 'Iris-setosa', 'Iris-setosa',<br>
>        'Iris-setosa', 'Iris-setosa', 'Iris-versicolor', 'Iris-versicolor',<br>
>        'Iris-versicolor', 'Iris-versicolor', 'Iris-versicolor',<br>
>        'Iris-versicolor', 'Iris-virginica', 'Iris-virginica',<br>
>        'Iris-virginica', 'Iris-virginica', 'Iris-virginica',<br>
>        'Iris-virginica', 'Iris-virginica', 'Iris-virginica'],
>       dtype='<U15')

> [0, 1, 2, 3, 4, 5, 0, 1, 2, 3, 4, 5, 0, 1, 2, 3, 4, 5, 6, 7]

53.Create group ids based on a given categorical variable. Use the following sample from iris species as input. Input:

url = 'https://archive.ics.uci.edu/ml/machine-learning-databases/iris/iris.data'

species = np.genfromtxt(url, delimiter=',', dtype='str', usecols=4)

species_small = np.sort(np.random.choice(species, size=20))

species_small
> array(['Iris-setosa', 'Iris-setosa', 'Iris-setosa', 'Iris-setosa',<br>
>        'Iris-setosa', 'Iris-setosa', 'Iris-versicolor', 'Iris-versicolor',<br>
>        'Iris-versicolor', 'Iris-versicolor', 'Iris-versicolor',<br>
>        'Iris-versicolor', 'Iris-virginica', 'Iris-virginica',<br>
>        'Iris-virginica', 'Iris-virginica', 'Iris-virginica',<br>
>        'Iris-virginica', 'Iris-virginica', 'Iris-virginica'],
>       dtype='<U15')

> [0, 0, 0, 0, 0, 1, 1, 1, 1, 1, 1, 1, 1, 1, 2, 2, 2, 2, 2, 2]

54.Create the ranks for the given numeric array a. Input:

np.random.seed(10)

a = np.random.randint(20, size=10)

print(a)
> [ 9  4 15  0 17 16 17  8  9  0]

[4 2 6 0 8 7 9 3 5 1]

55.Create a rank array of the same shape as a given numeric array a. Input:

np.random.seed(10)

a = np.random.randint(20, size=[2,5])

print(a)
> [[ 9  4 15  0 17]<br>
>  [16 17  8  9  0]]

> [[4 2 6 0 8]<br>
>  [7 9 3 5 1]]

56.Compute the maximum for each row in the given array.

np.random.seed(100)
a = np.random.randint(1,10, [5,3])
a
> array([[9, 9, 4],<br>
>        [8, 8, 1],<br>
>        [5, 3, 6],<br>
>        [3, 3, 3],<br>
>        [2, 1, 9]])



> array([9, 8, 6, 3, 9])

57.Compute the min-by-max for each row for given 2d numpy array.

np.random.seed(100)
a = np.random.randint(1,10, [5,3])
a
> array([[9, 9, 4],<br>
>        [8, 8, 1],<br>
>        [5, 3, 6],<br>
>        [3, 3, 3],<br>
>        [2, 1, 9]])<br>

> array([ 0.44444444,  0.125     ,  0.5       ,  1.        ,  0.11111111])

58.ind the duplicate entries (2nd occurrence onwards) in the given numpy array and mark them as True. First time occurrences should be False.

 Input<br>
np.random.seed(100)

a = np.random.randint(0, 5, 10)

print('Array: ', a)<br>
> Array: [0 0 3 0 2 4 2 2 2 2]

> [False  True False  True False False  True  True  True  True]

59.Find the mean of a numeric column grouped by a categorical column in a 2D numpy array Input:

url = 'https://archive.ics.uci.edu/ml/machine-learning-databases/iris/iris.data'

iris = np.genfromtxt(url, delimiter=',', dtype='object')

names = ('sepallength', 'sepalwidth', 'petallength', 'petalwidth', 'species')

> [[b'Iris-setosa', 3.418],<br>
>  [b'Iris-versicolor', 2.770],<br>
>  [b'Iris-virginica', 2.974]]

60.Import the image from the following URL and convert it to a numpy array. URL = ‘https://upload.wikimedia.org/wikipedia/commons/8/8b/Denali_Mt_McKinley.jpg

61.Drop all nan values from a 1D numpy array
Input:

np.array([1,2,3,np.nan,5,6,7,np.nan])

array([ 1.,  2.,  3.,  5.,  6.,  7.])

62.Compute the euclidean distance between two arrays a and b. Input:

a = np.array([1,2,3,4,5])

b = np.array([4,5,6,7,8])

> 6.7082039324993694

63.Find all the peaks in a 1D numpy array a. Peaks are points surrounded by smaller values on both sides.

 Input:

 a = np.array([1, 3, 7, 1, 2, 6, 0, 1]

> array([2, 5])<br>
where, 2 and 5 are the positions of peak values 7 and 6.

64.Subtract the 1d array b_1d from the 2d array a_2d, such that each item of b_1d subtracts from respective row of a_2d.

a_2d = np.array([[3,3,3],[4,4,4],[5,5,5]])

b_1d = np.array([1,1,1]

> [[2 2 2]<br>  [2 2 2]
>  <br>[2 2 2]]

65.Find the index of 5th repetition of number 1 in x.

x = np.array([1, 2, 1, 1, 3, 4, 3, 1, 1, 2, 1, 1, 2])

> 8

66.Convert numpy’s datetime64 object to datetime’s datetime object

 Input: a numpy datetime64 object dt64 = np.datetime64('2018-02-25 22:10:10')

> datetime.datetime(2018, 2, 25, 22, 10, 10)

67.Compute the moving average of window size 3, for the given 1D array. Input:

np.random.seed(100)

Z = np.random.randint(10, size=10)

> array:  [8 8 3 7 7 0 4 2 5 2]<br>
> moving average:  [ 6.33  6.    5.67  4.67  3.67  2.    3.67  3.  ]

68.Create a numpy array of length 10, starting from 5 and has a step of 3 between consecutive numbers

> array([ 5,  8, 11, 14, 17, 20, 23, 26, 29, 32])

69.Given an array of a non-continuous sequence of dates. Make it a continuous sequence of dates, by filling in the missing dates. Input:

 Input<br>
dates = np.arange(np.datetime64('2018-02-01'), np.datetime64('2018-02-25'), 2)

print(dates)
> ['2018-02-01' '2018-02-03' '2018-02-05' '2018-02-07' '2018-02-09'<br>
>  '2018-02-11' '2018-02-13' '2018-02-15' '2018-02-17' '2018-02-19'<br>
>  '2018-02-21' '2018-02-23']

> array(['2018-02-01', '2018-02-02', '2018-02-03', '2018-02-04',<br>
>        '2018-02-05', '2018-02-06', '2018-02-07', '2018-02-08',<br>
>        '2018-02-09', '2018-02-10', '2018-02-11', '2018-02-12',<br>
>        '2018-02-13', '2018-02-14', '2018-02-15', '2018-02-16',<br>
>        '2018-02-17', '2018-02-18', '2018-02-19', '2018-02-20',<br>
>        '2018-02-21', '2018-02-22', '2018-02-23'], dtype='datetime64[D]')

70.From the given 1d array arr, generate a 2d matrix using strides, with a window length of 4 and strides of 2, like [[0,1,2,3], [2,3,4,5], [4,5,6,7]..] Input:

arr = np.arange(15)

arr<br>
> array([ 0,  1,  2,  3,  4,  5,  6,  7,  8,  9, 10, 11, 12, 13, 14])

> [[ 0  1  2  3]<br>
>  [ 2  3  4  5]<br>
>  [ 4  5  6  7]<br>
>  [ 6  7  8  9]<br>
>  [ 8  9 10 11]<br>
>  [10 11 12 13]]